lib/Conversion/ExportAIGER/ExportAIGER.cpp

```diff
@@ -0,0 +1,739 @@
+    if (failed(visit(&op)))
+      return failure();
+  }
+
+  LLVM_DEBUG(llvm::dbgs() << "Found " << getNumAnds() << " AND gates, "
+                          << getNumLatches() << " latches\n");
+  return success();
+}
+
+LogicalResult AIGERExporter::visit(Operation *op) {
+  return TypeSwitch<Operation *, LogicalResult>(op)
+      .Case<aig::AndInverterOp>([&](auto op) { return visit(op); })
+      .Case<seq::CompRegOp>([&](auto op) { return visit(op); })
+      .Case<comb::ConcatOp>([&](auto op) { return visit(op); })
+      .Case<comb::ExtractOp>([&](auto op) { return visit(op); })
+      .Case<comb::ReplicateOp>([&](auto op) { return visit(op); })
+      .Case<hw::HWModuleOp, hw::ConstantOp, hw::OutputOp>(
+          [](auto) { return success(); })
+      .Default([&](Operation *op) -> LogicalResult {
+        // Handle unknown operations if handler is set
+        if (!handler)
+          return mlir::emitError(op->getLoc(), "unhandled operation: ") << *op;
+
+        // Handle unknown operations through the handler
+        for (mlir::OpOperand &operand : op->getOpOperands()) {
+          if (operand.get().getType().isInteger())
+            for (int64_t i = 0; i < getBitWidth(operand.get()); ++i) {
+              if (handler->operandCallback(operand, i, outputs.size()))
+                addOutput({operand.get(), i});
+            }
+        }
+
+        for (mlir::OpResult result : op->getOpResults()) {
+          if (!result.getType().isInteger())
+            continue;
+          for (int64_t i = 0; i < getBitWidth(result); ++i) {
+            if (handler->valueCallback(result, i, inputs.size()))
+              addInput({result, i});
+            else
+              // Treat it as a constant
+              valueLiteralMap[{result, i}] = 0;
+          }
+        }
+
+        return success();
+      });
+}
+
+LogicalResult AIGERExporter::visit(aig::AndInverterOp op) {
+  // Handle AIG AND-inverter operations
+  if (op.getInputs().size() == 1) {
+    // Single input = inverter (not counted as AND gate in AIGER)
+    // Inversion flag is looked at in getLiteral()
+    LLVM_DEBUG(llvm::dbgs() << "  Found inverter: " << op << "\n");
+  } else if (op.getInputs().size() == 2) {
+    // Two inputs = AND gate
+    for (int64_t i = 0; i < getBitWidth(op); ++i)
+      andGates.push_back({{op.getResult(), i},
+                          {op.getInputs()[0], i},
+                          {op.getInputs()[1], i}});
+
+    LLVM_DEBUG(llvm::dbgs() << "  Found AND gate: " << op << "\n");
+  } else {
+    // Variadic AND gates need to be lowered first
+    return mlir::emitError(op->getLoc(),
+                           "variadic AND gates not supported, run "
+                           "aig-lower-variadic pass first");
+  }
+  if (handler)
+    handler->notifyEmitted(op);
+  return success();
+}
+
+LogicalResult AIGERExporter::visit(seq::CompRegOp op) {
+  if (!op.getName())
+    op.setName(nameGenerator.newName("GEN_REG"));
```

Comment:
Does the AIGER format require registers to carry a name? If not, it might not be necessary to modify the IR at all :thinking: